src/fluree/db/query/range.cljc

```diff
@@ -56,154 +50,234 @@
-                                                    acci
-                                                    (disj acci f)))))))
-               i*           (count acc*)
-               more?        (and rhs
-                                 (neg? (idx-compare rhs end-flake))
-                                 (< i* limit))]
-           (if-not more?
-             acc*
-             (recur (<? (dbproto/-lookup-leaf root-node rhs)) i* acc*))))))))
-
-
-(defn subject-groups->allow-flakes
-  "Starting with flakes grouped by subject id, filters the flakes until
-  either flake-limit or subject-limit reached."
-  [db subject-groups flake-start subject-start flake-limit subject-limit]
-  (async/go
-    (loop [[subject-flakes & r] subject-groups
-           flake-count   flake-start
-           subject-count subject-start
-           acc           []]
-      (if (or (nil? subject-flakes) (>= flake-count flake-limit) (>= subject-count subject-limit))
-        [flake-count subject-count acc]
-        (let [subject-filtered #?(:clj (<? (perm-validate/allow-flakes? db subject-flakes))
-                                  :cljs subject-flakes)
-              flakes-new-count         (count subject-filtered)
-              subject-new-count        (if (= 0 flakes-new-count) 0 1)]
-          (recur r (+ flake-count flakes-new-count)
-                 (+ subject-count subject-new-count)
-                 (into acc subject-filtered)))))))
-
-(defn find-next-valid-node
-  [root-node rhs t novelty fast-forward-db?]
-  (go-try
-    (loop [lookup-leaf (<? (dbproto/-lookup-leaf root-node rhs))]
-      (let [node (try*
-                   (<? (dbproto/-resolve-to-t lookup-leaf t novelty fast-forward-db?))
-                   (catch* e nil))]
-        (if node node
-                 (if-let [rhs (:rhs lookup-leaf)]
-                   (recur (<? (dbproto/-lookup-leaf root-node rhs)))
-                   nil))))))
+  ([{t :t :as db} idx start-test start-match end-test end-match opts]
+   (let [{:keys [limit from-t to-t]
+          :or   {from-t t}}
+         opts
+
+         novelty     (get-in db [:novelty idx])
+         idx-compare (get-in db [:index-configs idx :comparator])
+         start-parts (match->flake-parts db idx start-match)
+         end-parts   (match->flake-parts db idx end-match)
+
+         out-chan    (chan 1 (map (fn [flakes]
+                                    (apply flake/sorted-set-by idx-compare flakes))))]
+     (go
+       (let [start-flake (<? (resolve-match-flake db start-test start-parts))
+             end-flake   (<? (resolve-match-flake db end-test end-parts))]
+         (-> db
+             (index-node-stream idx start-flake end-flake)
+             (expand-history-range from-t to-t novelty start-test start-flake
+                                   end-test end-flake)
+             (filter-authorized db start-flake end-flake)
+             (take-only limit)
+             (->> (async/into #{}))
```

Comment:
I think out-chan is already going to put into a set, might be more efficient to put into a list or vector. Set placement will require individual equality checks - which get done twice here.